server/apps/microservices/src/processors/video-transcode.processor.ts

```diff
@@ -0,0 +1,57 @@
+import { Process, Processor } from '@nestjs/bull';
+import { Logger } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Job } from 'bull';
+import ffmpeg from 'fluent-ffmpeg';
+import { existsSync, mkdirSync } from 'fs';
+import { Repository } from 'typeorm';
+import { AssetEntity } from '../../../../libs/database/src/entities/asset.entity';
+import { APP_UPLOAD_LOCATION } from '../../../immich/src/constants/upload_location.constant';
+
+@Processor('video-conversion-queue')
+export class VideoTranscodeProcessor {
+  constructor(
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+  ) {}
+
+  @Process({ name: 'mp4-conversion', concurrency: 1 })
+  async mp4Conversion(job: Job) {
+    const { asset }: { asset: AssetEntity } = job.data;
+
+    if (asset.mimeType != 'video/mp4') {
+      const basePath = APP_UPLOAD_LOCATION;
+      const encodedVideoPath = `${basePath}/${asset.userId}/encoded-video`;
+
+      if (!existsSync(encodedVideoPath)) {
+        mkdirSync(encodedVideoPath, { recursive: true });
+      }
+
+      const savedEncodedPath = encodedVideoPath + '/' + asset.id + '.mp4';
+
+      if (asset.encodedVideoPath == '' || !asset.encodedVideoPath) {
+        // Put the processing into its own async function to prevent the job exist right away
+        await this.runFFMPEGPipeLine(asset, savedEncodedPath);
+      }
+    }
+  }
+
+  async runFFMPEGPipeLine(asset: AssetEntity, savedEncodedPath: string): Promise<void> {
```

Comment:
This job at first didn't conform to the concurrency setting due to the callback functions of the FFmpeg command. I have to refactor to a separate async function so that the `process` function return after the job is either success or failed